File: api/src/models/booking.js
// models/booking.model.js
const mongoose = require("mongoose");

const bookingSchema = new mongoose.Schema(
  {
    user_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
    showtime_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "ShowTime",
      required: true,
    },
    total_seat: { type: Number, min: 0, required: true },
    status: {
      type: String,
      enum: ["PENDING", "CANCELLED", "COMPLETED"],
      default: "PENDING",
      required: true,
    },
    total_price_movie: { type: Number, min: 0, required: true },
    booking_seats: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "BookingSeat",
      },
    ],
    payment_id: { type: mongoose.Schema.Types.ObjectId, ref: "Payment" },
    notes: { type: String, maxlength: 500 },
  },
  { timestamps: { createdAt: "created_at", updatedAt: "updated_at" } }
);

// Index for better query performance
bookingSchema.index({ user_id: 1, status: 1 });
bookingSchema.index({ showtime_id: 1 });
bookingSchema.index({ created_at: -1 });

module.exports = mongoose.model("Booking", bookingSchema);

File: src/main/java/com/ra/base_spring_boot/model/Booking.java
package com.ra.base_spring_boot.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.ra.base_spring_boot.model.base.BaseObject;
import com.ra.base_spring_boot.model.constants.BookingStatus;
import jakarta.persistence.*;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.Min;
import lombok.*;

import java.util.Date;
import java.util.Set;

@Entity
@Table(name = "bookings")
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Builder
public class Booking extends BaseObject {
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", referencedColumnName = "id")
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "showtime_id", referencedColumnName = "id")
    private Showtime showtime;

    @Min(value = 0, message = "Total number of seats must not be less than 0")
    @Column(name = "total_seat")
    private int totalSeat;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private BookingStatus status;

    @DecimalMin(value = "0.0", inclusive = true, message = "Total amount cannot be less than 0")
    @Column(name = "total_price_movie")
    private double totalPriceMovie;

    @Column(name = "created_at")
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX", timezone = "Asia/Ho_Chi_Minh")
    private Date createdAt;

    @Column(name = "updated_at")
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX", timezone = "Asia/Ho_Chi_Minh")
    private Date updatedAt;

    @OneToMany(mappedBy = "booking", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<BookingSeat> bookingSeats;

    @OneToOne(mappedBy = "booking", cascade = CascadeType.ALL, orphanRemoval = true)
    private Payment payment;

    @PrePersist
    protected void onCreate() {
        Date now = new Date();
        this.createdAt = now;
        this.updatedAt = now;
        this.status = BookingStatus.PENDING;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = new Date();
    }
}